Sources/NIOPosix/Bootstrap.swift

```diff
@@ -482,6 +483,448 @@ public final class ServerBootstrap {
+            serverBackpressureStrategy: serverBackpressureStrategy,
+            childBackpressureStrategy: childBackpressureStrategy,
+            isChildChannelOutboundHalfClosureEnabled: isChildChannelOutboundHalfClosureEnabled
+        ) { (eventLoop, serverChannel) in
+            serverChannel.registerAndDoSynchronously { serverChannel in
+                serverChannel.bind(to: address)
+            }
+        }
+    }
+
+    private typealias MakeServerChannel = (_ eventLoop: SelectableEventLoop, _ childGroup: EventLoopGroup, _ enableMPTCP: Bool) throws -> ServerSocketChannel
+    private typealias Register = (EventLoop, ServerSocketChannel) -> EventLoopFuture<Void>
+
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    private func bindAsyncChannel0<ChildChannelInbound: Sendable, ChildChannelOutbound: Sendable>(
+        makeServerChannel: MakeServerChannel,
+        serverBackpressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark?,
+        childBackpressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark?,
+        isChildChannelOutboundHalfClosureEnabled: Bool,
+        _ register: @escaping Register
+    ) async throws -> NIOAsyncChannel<NIOAsyncChannel<ChildChannelInbound, ChildChannelOutbound>, Never>  {
+        let eventLoop = self.group.next()
+        let childEventLoopGroup = self.childGroup
+        let serverChannelOptions = self._serverChannelOptions
+        let serverChannelInit = self.serverChannelInit ?? { _ in eventLoop.makeSucceededFuture(()) }
+        let childChannelInit = self.childChannelInit
+        let childChannelOptions = self._childChannelOptions
+        
+        let serverChannel = try makeServerChannel(eventLoop as! SelectableEventLoop, childEventLoopGroup, self.enableMPTCP)
+        
+        return try await eventLoop.submit {
+            serverChannelOptions.applyAllChannelOptions(to: serverChannel).flatMap {
+                serverChannelInit(serverChannel)
+            }.flatMap {
+                do {
+                    try serverChannel.pipeline.syncOperations.addHandler(
+                        AcceptHandler(childChannelInitializer: childChannelInit, childChannelOptions: childChannelOptions),
+                        name: "AcceptHandler"
+                    )
+
+                    // We are wrapping the inbound channels into `NIOAsyncChannel`s with the transformation
+                    // closure of the `NIOAsyncChannel` that allows us to wrap them without adding
+                    // wrapping/unwrapping handlers to the pipeline.
+                    let asyncChannel = try NIOAsyncChannel<NIOAsyncChannel<ChildChannelInbound, ChildChannelOutbound>, Never>(
+                        synchronouslyWrapping: serverChannel,
+                        backpressureStrategy: serverBackpressureStrategy,
+                        transformationClosure: { channel in
+                            try NIOAsyncChannel(
```

Comment:
It appears this is being called on the server channel eventloop, but it expects to be called child channel eventloop.

Reply:
Thanks. Good catch!

Reply:
You should run your tests with an EventLoopGroup with more than one thread or at least have one test with this. Three is a good number to ensure you don't have a server and a client interleaving EventLoop usage. I was running with two threads and only caught this because I had a test which had two client requests in it which opened two child channels

Reply:
This is a very good piece of advice. I also recommend adding `preconditionInEventLoop` checks for any code that expects it, because they're fairly cheap and they flush out these bugs really early.